README.md
# About

**Status Dashboard** is frontend to consume the data provided by [Uptime Robot](https://uptimerobot.com/about)'s API.

The project is hosted on Github Pages, so it "never" gets down. The api keys in the project are read-only, so it's not a problem letting them in the repository.

# Data source

The data is provided by Uptime Robot's API through read-only API keys. Each read-only key has access to a single monitor.

To avoid making as many requests as the number of monitors enabled, I created a wrapper, hosted on my server, with a full access key that makes a single request for all monitors at once.

The wrapper's source-code is available at https://github.com/afonsodemori/aio-status-api. In case this endpoint is down, the frontend falls back to Uptime Robot's API.

# Links

  - Frontend: https://status.afonso.io
  - Frontend source: *(this repo)*
  - API: https://api.status.afonso.io
  - API source:  https://github.com/afonsodemori/aio-status-api


History.md
# 1.4.5, 2019-03-04
  - Add api key for srv-ops, srv-rpi, and rename home.afonso.io to srv-rpi.afonso.io

# 1.4.4, 2019-03-03
  - Add api key for home.afonso.io

# 1.4.3, 2019-02-28
  - [bugfix] Fix day count
  - Increase ajax timeout

# 1.4.2, 2019-02-25
  - Add api keys for api.sophia.afonso.io and pma.afonso.io

# 1.4.1, 2019-02-25
  - Add api key for api.status.afonso.io

# 1.4.0, 2019-02-22
  - [feature] Call own API to reduce number of requests; fallback to UptimeRobot's API if own API is unavailable
  - [bugfix] Fix wrong day range count
  - Add page for error 404
  - Remove integrity check from js and css

# 1.3.3, 2019-02-19
  - Remove logo and text-shadow from header
  - Change last 24h column to last 7d
  - Avoid summary waiting to have .fixed to change the status color
  - Change icons for not-checked-yet, seems-down and down
  - Change colors for monitor paused and not checked yet
  - Set timeout for ajax api requests
  - Fix theme color for about.html; change theme color in app.manifest

# 1.3.2, 2019-02-18
  - [hotfix] Fix assets URLs in service-worker.js

# 1.3.1, 2019-02-18
  - Add navbar
  - Create /about.html page and README.md
  - Fix mobile theme color for monitor-up

# 1.3.0, 2019-02-17
  - [feature] Add offline support
  - [bugfix] Fix count of paused/unknown monitors
  - Add notification when ajax request is not successful
  - Update favicons according to status
  - Fix data fetched from API, fix color coherence, add data columns
  - Keep the summary bar fixed when scrolling the page
  - Improve UX and change last downtime column to last status

# 1.2.0, 2019-02-16
  - [feature] New layout (add header, footer and UX improvements)
  - [feature] Add stats to the top of the page
  - [feature] Change color scheme according to the up/down status

# 1.1.0, 2019-02-15
  - [feature] Add column latest downtime
  - Add status.afonso.io to the list of services
  - Add countdown
  - Add viewport and improve mobile view
  - Add column for 15 days ratio
  - Improve UX on displaying uptime ratios
  - Add favicon.ico and robots.txt
  - Load api keys from external file
  - Create home template and fill it via jquery

# 1.0.0, 2019-02-14
  - Add basic implementation of status page


assets/scripts.js
const refresh = 15;

if (navigator.serviceWorker) {
    navigator.serviceWorker.register('/service-worker.js').then(function(){
        console.debug('SW: Register');
    }).catch(console.log);
}

function init(keys) {
    let append = '';
    for (let i = 0; i < keys.length; i++) {
        append += '<tr id="monitor-' + keys[i].id + '">' +
            '<th class="align-middle text-nowrap"><span class="status"></span> ' + keys[i].name + '</th>' +
            '<td class="align-middle latest">' +
            '<span class="reason">loading...</span>' +
            '<span class="time">loading...</span>' +
            '<span class="duration">loading...</span></td>'
        ;

        for (let r = 0; r < 9; r++) {
            append += '<td class="r-' + r + '"><span class="r' + r + ' ratio-color-loading">loading...</span></td>';
        }

        append += '</tr>';
    }

    $('table').append(append);

    loadData();
    checkStatus();
}

let apiKeys = null;
$.ajax({
    async: true,
    url: '/config/api-keys.json',
    dataType: 'json',
    success: function (data) {
        apiKeys = data;
        init(apiKeys);
    }
});

function formatDatetime(timestamp) {
    let date = new Date(timestamp * 1000);
    let options = {
        year: 'numeric',
        month: 'short',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit'
    };
    return new Intl.DateTimeFormat('en-GB', options).format(date);
}

function formatDuration(duration) {
    if (duration < 60) {
        return duration + ' sec';
    }

    if (duration < 60 * 60) {
        return Math.floor(duration / 60) + ' min, ' + duration % 60 + ' sec';
    }

    return Math.floor(duration / 60 / 60) + ' hour, ' +
        Math.floor((duration / 60) % 60) + ' min, ' +
        duration % 60 + ' sec';
}

function loadData() {
    let now = new Date();

    let last24h = (Math.floor(now.getTime() / 1000) - 60 * 60 * 24) + '_' + Math.floor(now.getTime() / 1000);
    let last30d = (Math.floor(now.getTime() / 1000) - 60 * 60 * 24 * 30) + '_' + Math.floor(now.getTime() / 1000);

    let ranges = last24h + '-' + last30d;

    let day = now;
    day.setHours(0, 0, 0, 0);
    let range = [];
    for (let i = 0; i < 7; i++) {
        $('.d-' + i).html(day.getDate() + '/' + (day.getMonth() + 1));
        range[0] = Math.floor(day.getTime() / 1000);
        range[1] = Math.floor(day.setDate(day.getDate() - 1) / 1000);
        ranges += '-' + range[1] + '_' + range[0];
    }

    for (let m = 0; m < apiKeys.length; m++) {
        $.ajax({
            type: 'post',
            url: 'https://api.uptimerobot.com/v2/getMonitors',
            data: {
                api_key: apiKeys[m].key,
                custom_uptime_ranges: ranges,
                logs: 1,
                logs_limit: 1
            },
            async: true,
            success: function (data) {
                let monitor = data.monitors[0];
                let id = '#monitor-' + monitor.id;
                let ratio = monitor.custom_uptime_ranges.split('-');

                $(id + ' .status')
                    .removeAttr('class')
                    .addClass('status')
                    .addClass('status-' + monitor.status)
                ;

                if (monitor.logs.length === 0) {
                    $(id + ' .latest .time').html('');
                    $(id + ' .latest .reason').html('');
                    $(id + ' .latest .duration').html('');
                }

                let log = monitor.logs[0];
                $(id + ' .latest .time').html(formatDatetime(log.datetime));
                $(id + ' .latest .reason').html(log.reason.code + ' - ' + log.reason.detail);
                $(id + ' .latest .duration').html(formatDuration(log.duration));

                let color = 'inherit';
                switch (monitor.status) {
                    case 0:
                        color = 'var(--monitor-paused)';
                        break;
                    case 1:
                        color = 'var(--monitor-not-checked-yet)';
                        break;
                    case 2:
                        color = 'var(--monitor-up)';
                        break;
                    case 8:
                        color = 'var(--monitor-seems-offline)';
                        break;
                    case 9:
                        color = 'var(--monitor-down)';
                        break;
                }

                $(id + ' .latest .reason').css('color', color);

                for (let i = 0; i < ratio.length; i++) {
                    let content = (ratio[i] === '100.000' ? 100 : ratio[i]) + '%';
                    let $el = $(id + ' .r' + i);
                    if (content !== $el.html()) {
                        $el
                            .removeAttr('class')
                            .addClass('r' + i)
                            .addClass('ratio-color-' + ratio[i].split('.')[0])
                            .html(content)
                            .hide()
                            .slideDown()
                        ;
                    }
                }

                $('.last-update').html(formatDatetime((new Date()).getTime() / 1000));
            },
            error: function (data) {
                let toast = '<div class="toast">Error fetching uptime data. Check your internet connection.<div>';
                if ($('.toast').length === 0) {
                    $('body').append(toast);
                    $('.toast')
                        .fadeIn()
                        .delay(5000)
                        .fadeOut(200, function () {
                            this.remove();
                        })
                    ;
                }
            }
        });
    }
}

function checkStatus() {
    let up = $('.status-2').length;
    let down = $('.status-9').length;
    let paused = $('.status-0').length;
    let unknown = $('.status').length - up - down - paused;

    $('.monitors-up').html(up);
    $('.monitors-down').html(down);
    $('.monitors-paused').html(paused);
    $('.monitors-unknown').html(unknown);

    if (down > 0) {
        $('header').css('background', 'var(--monitor-down)');
        $('.navbar').css('background-color', 'var(--monitor-down) !important');
        $('#summary.fixed').css('border-top-color', 'var(--monitor-down)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-down.ico');
        $('meta[name="theme-color"]').attr('content', '#D32F2F');
    } else if (up > 0) {
        $('header').css('background', 'var(--monitor-up)');
        $('.navbar').css('background-color', 'var(--monitor-up)');
        $('#summary.fixed').css('border-top-color', 'var(--monitor-up)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-up.ico');
        $('meta[name="theme-color"]').attr('content', '#7CB342');
    } else {
        $('header').css('background', 'var(--monitor-not-checked-yet)');
        $('.navbar').css('background-color', 'var(--monitor-not-checked-yet) !important');
        $('#summary.fixed').css('border-top-color', 'var(--monitor-not-checked-yet)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-not-checked-yet.ico');
        $('meta[name="theme-color"]').attr('content', '#E0E0E0');
    }
}

$(document).on('scroll', function () {
    let $div = $('#summary');
    let pos = $('header').height();
    if (pos < window.scrollY) {
        $div.addClass('fixed');
        $('body').css('padding-top', $div.height());
    } else {
        $('body').css('padding-top', 0);
        $div.removeClass('fixed');
    }
});

let timeout = refresh;
setInterval(function () {
    checkStatus();
    if (--timeout === 0) {
        loadData();
        $('.countdown').html(timeout);
        timeout = refresh;
    } else {
        $('.countdown').html(timeout);
    }
}, 1000);


service-worker.js
// version: 1.4.5

self.addEventListener('install', function (event) {
    event.waitUntil(
        caches.open('cache').then(function (cache) {
            return cache.addAll(
                [
                    '/',
                    '/about.html',

                    '/config/api-keys.json',

                    '/assets/css/general.css',
                    '/assets/css/dashboard.css',
                    '/assets/js/dashboard.js',

                    '/assets/app.manifest',

                    '/assets/img/favicons/android-chrome-192x192.png',
                    '/assets/img/favicons/android-chrome-512x512.png',

                    '/assets/img/favicons/favicon-down.ico',
                    '/assets/img/favicons/favicon-not-checked-yet.ico',
                    '/assets/img/favicons/favicon-seems-down.ico',
                    '/assets/img/favicons/favicon-up.ico',

                    'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css',
                    'https://fonts.googleapis.com/css?family=Open+Sans:300',
                    'https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js',
                    'https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js',
                    'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js',
                ]
            );
        })
    );
});

self.addEventListener('fetch', function (event) {
    event.respondWith(
        caches.open('cache').then(function (cache) {
            return cache.match(event.request).then(function (response) {
                return response || fetch(event.request).then(function (response) {
                    cache.put(event.request, response.clone());
                    return response;
                });
            });
        })
    );
});


assets/js/dashboard.js
const PAGE_REFRESH = 15;
const AJAX_TIMEOUT = 5;

if (navigator.serviceWorker) {
    navigator.serviceWorker.register('/service-worker.js').then(function(){
        console.debug('SW: Register');
    }).catch(console.log);
}

function init(keys) {
    let append = '';
    for (let i = 0; i < keys.length; i++) {
        append += '<tr id="monitor-' + keys[i].id + '">' +
            '<th class="align-middle text-nowrap"><span class="status"></span> ' + keys[i].name + '</th>' +
            '<td class="align-middle latest">' +
            '<span class="reason">loading...</span>' +
            '<span class="time">loading...</span>' +
            '<span class="duration">loading...</span></td>'
        ;

        for (let r = 0; r < 9; r++) {
            append += '<td class="r-' + r + '"><span class="r' + r + ' ratio-color-loading">loading...</span></td>';
        }

        append += '</tr>';
    }

    $('table').append(append);

    loadData();
    checkStatus();
}

let apiKeys = null;
$.ajax({
    url: '/config/api-keys.json',
    dataType: 'json',
    success: function (data) {
        apiKeys = data;
        init(apiKeys);
    }
});

function formatDatetime(timestamp) {
    let date = new Date(timestamp * 1000);
    let options = {
        year: 'numeric',
        month: 'short',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit'
    };
    return new Intl.DateTimeFormat('en-GB', options).format(date);
}

function formatDuration(duration) {
    if (duration < 60) {
        return duration + ' sec';
    }

    if (duration < 60 * 60) {
        return Math.floor(duration / 60) + ' min, ' + duration % 60 + ' sec';
    }

    return Math.floor(duration / 60 / 60) + ' hour, ' +
        Math.floor((duration / 60) % 60) + ' min, ' +
        duration % 60 + ' sec';
}

function loadData() {
    let now = new Date();

    let last7d = (Math.floor(now.getTime() / 1000) - 60 * 60 * 24 * 7) + '_' + Math.floor(now.getTime() / 1000);
    let last30d = (Math.floor(now.getTime() / 1000) - 60 * 60 * 24 * 30) + '_' + Math.floor(now.getTime() / 1000);

    let ranges = last7d + '-' + last30d;

    let day = now;
    day.setDate(day.getDate() + 1);
    day.setHours(0, 0, 0, 0);

    let range = [];
    for (let i = 0; i < 7; i++) {
        range[0] = Math.floor(day.getTime() / 1000);
        range[1] = Math.floor(day.setDate(day.getDate() - 1) / 1000);
        $('.d-' + i).html(day.getDate() + '/' + (day.getMonth() + 1));
        ranges += '-' + range[1] + '_' + range[0];
    }

    let monitors = [];
    for (let i = 0; i < apiKeys.length; i++) {
        monitors.push(apiKeys[i].id);
    }

    let ajaxSettings = {
        type: 'get',
        url: 'https://api.status.afonso.io/monitors',
        timeout: AJAX_TIMEOUT * 1000,
        async: true,
        data: {
            id: monitors.join('-'),
            custom_uptime_ranges: ranges,
            logs: 1,
            logs_limit: 1
        }
    };

    $.ajax(Object.assign(ajaxSettings, {
        success: function (data) {
            for (let i = 0; i < data.monitors.length; i++) {
                processMonitorData(data.monitors[i]);
            }
        },
        error: function (data) {
            delete ajaxSettings.data.id;
            loadFromUptimeRobot(ajaxSettings);
        }
    }));
}

function loadFromUptimeRobot(ajaxSettings) {
    ajaxSettings.type = 'post';
    ajaxSettings.url = 'https://api.uptimerobot.com/v2/getMonitors';

    for (let i = 0; i < apiKeys.length; i++) {
        ajaxSettings.data.api_key = apiKeys[i].key;
        $.ajax(Object.assign(ajaxSettings, {
            success: function (data) {
                processMonitorData(data.monitors[0]);
            },
            error: function (data) {
                let toast = '<div class="toast">Error fetching uptime data. Check your internet connection.<div>';
                if ($('.toast').length === 0) {
                    $('body').append(toast);
                    $('.toast')
                        .fadeIn()
                        .delay(5000)
                        .fadeOut(200, function () {
                            this.remove();
                        })
                    ;
                }
            }
        }));
    }
}

function processMonitorData(monitor) {
    let id = '#monitor-' + monitor.id;
    let ratio = monitor.custom_uptime_ranges.split('-');

    $(id + ' .status')
        .removeAttr('class')
        .addClass('status')
        .addClass('status-' + monitor.status)
    ;

    if (monitor.logs.length === 0) {
        $(id + ' .latest .time').html('');
        $(id + ' .latest .reason').html('');
        $(id + ' .latest .duration').html('');
    }

    let log = monitor.logs[0];
    $(id + ' .latest .time').html(formatDatetime(log.datetime));
    $(id + ' .latest .reason').html(log.reason.code + ' - ' + log.reason.detail);
    $(id + ' .latest .duration').html(formatDuration(log.duration));

    let color = 'inherit';
    switch (monitor.status) {
        case 0:
            color = 'var(--monitor-paused)';
            break;
        case 1:
            color = 'var(--monitor-not-checked-yet)';
            break;
        case 2:
            color = 'var(--monitor-up)';
            break;
        case 8:
            color = 'var(--monitor-seems-offline)';
            break;
        case 9:
            color = 'var(--monitor-down)';
            break;
    }

    $(id + ' .latest .reason').css('color', color);

    for (let i = 0; i < ratio.length; i++) {
        let content = (ratio[i] === '100.000' ? 100 : ratio[i]) + '%';
        let $el = $(id + ' .r' + i);
        if (content !== $el.html()) {
            $el
                .removeAttr('class')
                .addClass('r' + i)
                .addClass('ratio-color-' + ratio[i].split('.')[0])
                .html(content)
                .hide()
                .slideDown()
            ;
        }
    }

    $('.last-update').html(formatDatetime((new Date()).getTime() / 1000));
}

function checkStatus() {
    let up = $('.status-2').length;
    let down = $('.status-9').length;
    let paused = $('.status-0').length;
    let unknown = $('.status').length - up - down - paused;

    $('.monitors-up').html(up);
    $('.monitors-down').html(down);
    $('.monitors-paused').html(paused);
    $('.monitors-unknown').html(unknown);

    if (down > 0) {
        $('header').css('background', 'var(--monitor-down)');
        $('.navbar').css('background-color', 'var(--monitor-down)');
        $('#summary').css('border-top-color', 'var(--monitor-down)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-down.ico');
        $('meta[name="theme-color"]').attr('content', '#D32F2F');
    } else if (up > 0) {
        $('header').css('background', 'var(--monitor-up)');
        $('.navbar').css('background-color', 'var(--monitor-up)');
        $('#summary').css('border-top-color', 'var(--monitor-up)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-up.ico');
        $('meta[name="theme-color"]').attr('content', '#7CB342');
    } else {
        $('header').css('background', 'var(--monitor-not-checked-yet)');
        $('.navbar').css('background-color', 'var(--monitor-not-checked-yet)');
        $('#summary').css('border-top-color', 'var(--monitor-not-checked-yet)');
        $('link[rel="icon"]').attr('href', '/assets/img/favicons/favicon-not-checked-yet.ico');
        $('meta[name="theme-color"]').attr('content', '#BDBDBD');
    }
}

$(document).on('scroll', function () {
    let $div = $('#summary');
    let pos = $('header').height();
    if (pos < window.scrollY) {
        $div.addClass('fixed');
        $('body').css('padding-top', $div.height());
    } else {
        $('body').css('padding-top', 0);
        $div.removeClass('fixed');
    }
});

let timeout = PAGE_REFRESH;
setInterval(function () {
    checkStatus();
    if (--timeout === 0) {
        loadData();
        $('.countdown').html(timeout);
        timeout = PAGE_REFRESH;
    } else {
        $('.countdown').html(timeout);
    }
}, 1000);
